### ISSUE TITLE
Class constructor returns any (regression > 0.53.1)

### ISSUE THREAD

**Issue by decentralion** [facebook/flow#6400]
Consider the following code: [flow.org/try](https://flow.org/try/#0PTAEAEDMBsHsHcBQiDG0CGBnToBC6AnUAb0VFBVgDtMAXAgVxVtgIAoBKEgX2XIFtWAU3zsOALjyEA2gF0SZcqAJDaDAlVByA3It69EkBlWYBLaqFrxTAExvQhbAEaFJorqXKUatUIQLoAJ4AymqQkKAAvKAAggQBgQB0kASw-M6EiYIqopwcugbedKAAHlGgVELwUmK6JYmYaaoAFqZUAOYAcrC0ACJCkG1CNsHNsAzQNgCi8aycukA)
```js
// @flow

class Bar {
  constructor() {}

  moreBar(): Bar[] {
    return [];
  }
}

function twiddle(bar: Bar) {
  const arrayStuff = Array.from(bar.moreBar());
}

const x = new Bar();
x.somethingNotDefinedShouldError();
```

Clearly, it should error, as follows:
```
    16: x.somethingNotDefinedShouldError();
          ^ property `somethingNotDefinedShouldError`. Property not found in
    16: x.somethingNotDefinedShouldError();
        ^ Bar
```

And, in Flow 0.53.1 and previous, it did correctly error. However, as of version 54 and up, the code incorrectly passes.

The example may look a bit wonky, but it's as minimal as I can get it. Specifically, removing the empty constructor or the call to `Array.from` fixes the issue.

**Comment by wchargin** [comment]
Here is another way to reproduce this issue:

```js
// @flow

declare class Overload {
  static id<T>(x: T, nothing: void): T;
  // (the actual type of `nothing` doesn't matter)
  static id<T>(x: T): T;
}

class Bar {
  constructor() {}

  necessary(): Bar[] {
    // This method needs to be present, and needs to return a type that
    // "contains" `Bar` in some way: `Bar` and `Bar[]` and `Bar | 17`
    // are all fine, but `string` is not. The actual return value
    // doesn't seem to matter.
    return [];
  }
}

function twiddle(bar: Bar) {
  Overload.id(bar);
}

const x = new Bar();
x.somethingNotDefinedShouldError();
```

[(flow.org/try)][1]

Like the code in the OP, this code calls an overloaded method:
`Overload.id` has two types declared, and `Array.from` has six.

Unlike the code in the OP, this problem repros all the way back to
v0.37.0 (the earliest available version on flow/try).

Indeed, if the six `lib/core.js` definitions for `from` are replaced by

```js
    static from<A>(iter: Iterable<A>): Array<A>;
```

and Flow is rebuilt, then the problem is fixed—but of course lots of
other tests fail, because those other signatures are actually needed.

Commit cfa3ba7204574304c6700047f495acca7bba495c sounds relevant:

> Overloads + any sigs are bad -- even if a different signature matches
> on the params, if the return value is mis-used then Flow lets it fall
> into the "any" signature. This basically means any overload with an
> "any" fallback has almost no typechecking other than that fallback.

@samwgoldman: Looks like you reviewed that commit; do you have any
context that might be helpful?

(I tried building Flow at that commit to investigate, but the build
fails due to a missing `.cmi` file, even when I switch my OCaml compiler
to 4.01.0 (the README-recommended version at the time) and build from
scratch.)

[1]: https://flow.org/try/#0PTAEAEDMBsHsHcBQiAmBTAxtAhgJzaFtgM7GgDyAbmrnNiqAN6KijEAu27AlhqNygA8AFQB8ACgAeALlDCANKAB2sdgAtuSgOazKsAQEpZwgNwtQIUOPUFsGdgFds0UOwCeABwKxIoAAYq6ppafqAosGjESgDk7KAAtlzsNAbmHFy8-EJiUsZGcmYAvshEpKAAQnhM5hiwShy4DvawuOIGTMXmSpiRxHhubbKVuADaALrVrKyWwhpk8WjqsAzdaChk7LCgAEYEHvjEaErsithKK2hrG1v4jrhKoNiungTqXObTYABEtcfYmsQvv5hqFNGxYAtQPBsG5ZH4QY9zsC8ONQmcGPCqgAfUAARgA7H4PhYwHhbNAXJBNGhFNsHHE-A1gqCyIEAHRyNS2exOFy3Bz3UCUZwONDEyzhSIxOKHNDxVxbRLsZK4NnE-mC8ZmVjFTqQBxKezcOqueACFDQNDibZ4IZ4drMVhUGh0FBsgTW+1FEp1DigSSgAC8yjQ8AqeDaZkkbOIEMWGm0ADlVAARNBU7ooADKalgDmgKAAorhcC1I4ggA

**Comment by wchargin** [comment]
In either example, if the type of the constructor is annotated to return
`void`, the problem is fixed. In fact, the problem is fixed if the
constructor is annotated to have nearly any return type:

  - `Bar`, `void`, `null`, `string`, `number`, `{}`, and `mixed` all
    work, as does `Function` (which is initially a little surprising,
    but actually makes sense);
  - `any` and `Object` don’t work (which is not surprising);
  - `empty` does not work (which is somewhat surprising).

Similarly, if the class definition for `Bar` is replaced with a libdef

```js
declare class Bar {
  constructor(): void;  // this line may be omitted (same effect)
  moreBar(): Bar[];
}
```

then the problem is fixed.

Compounded with the fact that omitting the constructor in the definition
fixes the problem, it looks like this might have to do with inferring
the type of the constructor?

**Comment by wchargin** [comment]
Using the test case in the OP [with this setup][1], the first bad commit
is d4a23e383903fb322bf6a588c9de9e078c53dbf9.

[1]: https://gist.github.com/wchargin/dba8bc3a74c6e6ff12b466c187fdd4a5

It seems likely, though, that this commit only revealed an underlying
bug; it is not the root cause.

**Comment by stephenball1** [comment]
We just ran into this issue as well, so I'm glad a bug has already been filed.

Here's the minimal test case I found, in case it helps. Like the first example, the presence of an unused method causes the class instance to become `any`. Removing the unused method fixes the type checking.

```js
class Test<T> {
  value: T;
  constructor(value: T) {
    this.value = value;
  }
}

function create<T>(input: T | Test<T>): Test<T> {
  if (input instanceof Test) {
    return new Test(input.value);
  }
  return new Test(input);  
}

const instance = new Test(1);
(instance: boolean);
```
Marking the constructor's return type, as @wchargin suggested, fixes the problem here too.

**Comment by wchargin** [comment]
@stephenball1: Thanks. Unrelated to the main issue of this thread, your
code is unsound for a different reason: given your `create`, we may
write

```js
const t: Test<Test<string>> = create(new Test("hi"));
const s: string = t.value.value;  // actually undefined
s.toLowerCase();
```

which does (and should) type-check, but has a runtime error because the
runtime type of `t` is only `Test<string>`.

Your function `create: <T>(T | Test<T>) => Test<T>` is in fact
unimplementable, because the two calls to `create` in

```js
const base: Test<string> = new Test("hi");
const same: Test<string> = create<string>(base);
const next: Test<Test<string>> = create(base);
```

cannot be distinguished at runtime and are nevertheless required to
return different results.

The root problem is that the refinement `input instanceof Test` is
intended to refine the type of `input: T | Test<T>` to `Test<T>`, but
this is of course not valid when `T` is itself a `Test`. It is
unforunate that Flow does not detect this.

Anyway, like I said, unrelated; just thought that you’d want to know.

**Comment by stephenball1** [comment]
Ah great point. Thanks for the explanation @wchargin, it makes a lot of sense.

I originally had a more concrete type `value: Array<T>` inside `Test`, with `create: <T>(Array<T> | Test<T>) => Test<T>`, but removed the `Array`  requirement to see if the bug still occurred.

**Comment by wchargin** [comment]
Yep, that should work; the type constructors are both distinct and
present at runtime (as opposed to, e.g., type aliases), so you should be
able to disambiguate.

**Comment by roryokane** [comment]
I ran into this issue too in [#6738 “Object construction is wrongly inferred to return `any` when constructing a class with a Promise-using method”][my-issue]. This was my minimal test case (explained in more detail in that issue):

~~~javascript
// @flow

class Engine {
  constructor() {}
  
  pipe(aPromise: Promise<null>): Engine {
    aPromise.then(() => this);
    return this;
  }
}

const anEngine = new Engine(); // bug: `anEngine` has type `any` instead of `Engine`
~~~

[my-issue]: https://github.com/facebook/flow/issues/6738

**Comment by wchargin** [comment]
Fixed in Flow v0.98.0.
